Failing to retrieve service type, brings node down

**Description**
The node aborts during repeated "Failed to retrieve service type" and crashes the node.


```
SUMMARY
========

PARAMETERS
 * /foxglove_bridge/address: 0.0.0.0
2023-09-01T21:43:49.718834923Z  * /foxglove_bridge/asset_uri_allowlist: ['package://(/?\\...
2023-09-01T21:43:49.718837287Z  * /foxglove_bridge/capabilities: ['clientPublish',...
 * /foxglove_bridge/certfile: 
 * /foxglove_bridge/client_topic_whitelist: ['.*']
2023-09-01T21:43:49.718846455Z  * /foxglove_bridge/keyfile: 
 * /foxglove_bridge/max_update_ms: 5000
 * /foxglove_bridge/param_whitelist: ['.*']
 * /foxglove_bridge/port: 8765
 * /foxglove_bridge/send_buffer_limit: 10000000
 * /foxglove_bridge/service_whitelist: ['.*']
 * /foxglove_bridge/tls: False
2023-09-01T21:43:49.718861954Z  * /foxglove_bridge/topic_whitelist: ['.*']
 * /foxglove_nodelet_manager/num_worker_threads: 0
 * /rosdistro: noetic
2023-09-01T21:43:49.718868607Z  * /rosversion: 1.16.0

NODES
  /
    foxglove_bridge (nodelet/nodelet)
    foxglove_nodelet_manager (nodelet/nodelet)

ROS_MASTER_URI=http://hood1-1-rios-roscore-service:80/
]2;/opt/rios/install/share/foxglove_bridge/launch/foxglove_bridge.launch http://hood1-1-rios-roscore-service:80/
process[foxglove_nodelet_manager-1]: started with pid [68]
process[foxglove_bridge-2]: started with pid [69]
2023-09-01T21:43:49.954738356Z [ERROR] [1693604629.954637736]: Failed to retrieve service type or service description of service /rviz_msiraider_455523_4051231953047898198/get_loggers: std::future_error: Broken promise
[ERROR] [1693604629.960421416]: Failed to retrieve service type or service description of service /rviz_msiraider_455523_4051231953047898198/set_logger_level: std::future_error: Broken promise
[ERROR] [1693604629.963452575]: Failed to retrieve service type or service description of service /rviz_msiraider_455523_4051231953047898198/reload_shaders: std::future_error: Broken promise
[ERROR] [1693604629.966846903]: Failed to retrieve service type or service description of service /rviz_msiraider_455523_4051231953047898198/load_config: std::future_error: Broken promise
[ERROR] [1693604629.970424157]: Failed to retrieve service type or service description of service /rviz_msiraider_455523_4051231953047898198/load_config_discarding_changes: std::future_error: Broken promise
[ERROR] [1693604629.974170433]: Failed to retrieve service type or service description of service /rviz_msiraider_455523_4051231953047898198/save_config: std::future_error: Broken promise
[ERROR] [1693604629.976027506]: Failed to retrieve service type or service description of service /rqt_gui_cpp_node_3355783/get_loggers: std::future_error: Broken promise
[ERROR] [1693604629.977745424]: Failed to retrieve service type or service description of service /rqt_gui_cpp_node_3355783/set_logger_level: std::future_error: Broken promise
[ERROR] [1693604629.982635479]: Failed to retrieve service type or service description of service /rqt_gui_cpp_node_3360882/get_loggers: std::future_error: Broken promise
[ERROR] [1693604629.985879822]: Failed to retrieve service type or service description of service /rqt_gui_cpp_node_3360882/set_logger_level: std::future_error: Broken promise
[ERROR] [1693604629.995735004]: Failed to retrieve service type or service description of service /tf_echo_1693005407704472780/get_loggers: std::future_error: Broken promise
[ERROR] [1693604629.997922693]: Failed to retrieve service type or service description of service /tf_echo_1693005407704472780/set_logger_level: std::future_error: Broken promise
[ERROR] [1693604630.002444470]: Failed to retrieve service type or service description of service /rviz_msiraider_475863_1805560526383919607/get_loggers: std::future_error: Broken promise
[ERROR] [1693604630.007448601]: Failed to retrieve service type or service description of service /rviz_msiraider_475863_1805560526383919607/set_logger_level: std::future_error: Broken promise
2023-09-01T21:43:50.013426630Z [ERROR] [1693604630.013370403]: Failed to retrieve service type or service description of service /rviz_msiraider_475863_1805560526383919607/reload_shaders: std::future_error: Broken promise
[ERROR] [1693604630.018472310]: Failed to retrieve service type or service description of service /rviz_msiraider_475863_1805560526383919607/load_config: std::future_error: Broken promise
[ERROR] [1693604630.023164429]: Failed to retrieve service type or service description of service /rviz_msiraider_475863_1805560526383919607/load_config_discarding_changes: std::future_error: Broken promise
[ERROR] [1693604630.028526369]: Failed to retrieve service type or service description of service /rviz_msiraider_475863_1805560526383919607/save_config: std::future_error: Broken promise
[ERROR] [1693604630.033903217]: Failed to retrieve service type or service description of service /rqt_gui_cpp_node_68845/get_loggers: std::future_error: Broken promise
[ERROR] [1693604630.038839359]: Failed to retrieve service type or service description of service /rqt_gui_cpp_node_68845/set_logger_level: std::future_error: Broken promise
[ WARN] [1693604630.291370231]: Failed to retrieve type of service /rqt_gui_cpp_node_11371/get_loggers
2023-09-01T21:43:50.543664253Z [ WARN] [1693604630.543572709]: Failed to retrieve type of service /rqt_gui_cpp_node_11371/set_logger_level
[ WARN] [1693604630.795988773]: Failed to retrieve type of service /rqt_gui_cpp_node_15179/get_loggers
[ WARN] [1693604631.049000466]: Failed to retrieve type of service /rqt_gui_cpp_node_15179/set_logger_level
[ WARN] [1693604631.302035504]: Failed to retrieve type of service /rqt_gui_cpp_node_15179/compressed/set_parameters
[ERROR] [1693604631.305001208]: Failed to retrieve service type or service description of service /tf_echo_1693329846940349580/get_loggers: std::future_error: Broken promise
[ERROR] [1693604631.307799466]: Failed to retrieve service type or service description of service /tf_echo_1693329846940349580/set_logger_level: std::future_error: Broken promise
[ERROR] [1693604631.312068543]: Failed to retrieve service type or service description of service /rviz_msiraider_613058_4222886344543863053/get_loggers: std::future_error: Broken promise
[ERROR] [1693604631.315257622]: Failed to retrieve service type or service description of service /rviz_msiraider_613058_4222886344543863053/set_logger_level: std::future_error: Broken promise
2023-09-01T21:43:51.319103386Z [ERROR] [1693604631.318751839]: Failed to retrieve service type or service description of service /rviz_msiraider_613058_4222886344543863053/reload_shaders: std::future_error: Broken promise
[ERROR] [1693604631.322952897]: Failed to retrieve service type or service description of service /rviz_msiraider_613058_4222886344543863053/load_config: std::future_error: Broken promise
[ERROR] [1693604631.327608327]: Failed to retrieve service type or service description of service /rviz_msiraider_613058_4222886344543863053/load_config_discarding_changes: std::future_error: Broken promise
[ERROR] [1693604631.331009648]: Failed to retrieve service type or service description of service /rviz_msiraider_613058_4222886344543863053/save_config: std::future_error: Broken promise
[ERROR] [1693604631.379363711]: Failed to retrieve service type or service description of service /rostopic_614368_1693332654733/get_loggers: std::future_error: Broken promise
[ERROR] [1693604631.383728559]: Failed to retrieve service type or service description of service /rostopic_614368_1693332654733/set_logger_level: std::future_error: Broken promise
[ERROR] [1693604631.386775559]: Failed to retrieve service type or service description of service /rviz_msiraider_618543_7783352244266021340/get_loggers: std::future_error: Broken promise
[ERROR] [1693604631.390348856]: Failed to retrieve service type or service description of service /rviz_msiraider_618543_7783352244266021340/set_logger_level: std::future_error: Broken promise
[ERROR] [1693604631.393720982]: Failed to retrieve service type or service description of service /rviz_msiraider_618543_7783352244266021340/reload_shaders: std::future_error: Broken promise
[ERROR] [1693604631.397023165]: Failed to retrieve service type or service description of service /rviz_msiraider_618543_7783352244266021340/load_config: std::future_error: Broken promise
[ERROR] [1693604631.400547830]: Failed to retrieve service type or service description of service /rviz_msiraider_618543_7783352244266021340/load_config_discarding_changes: std::future_error: Broken promise
[ WARN] [1693604633.884170701]: Could not find package 'camera_control_msgs' in srv paths!
[ERROR] [1693604633.884195017]: Failed to retrieve type information for service '/qc_camera1/get_cam_properties' of type 'camera_control_msgs/GetCamProperties'
[ WARN] [1693604633.887191851]: Could not find package 'arena_camera_ros' in srv paths!
2023-09-01T21:43:53.887251244Z [ERROR] [1693604633.887212631]: Failed to retrieve type information for service '/qc_camera1/toggle_stream' of type 'arena_camera_ros/ToggleCameraStream'
[ WARN] [1693604633.917033031]: Could not find package 'camera_control_msgs' in srv paths!
[ERROR] [1693604633.917065743]: Failed to retrieve type information for service '/qc_camera2/set_binning' of type 'camera_control_msgs/SetBinning'
2023-09-01T21:43:53.920068057Z [ WARN] [1693604633.920025937]: Could not find package 'camera_control_msgs' in srv paths!
[ERROR] [1693604633.920049983]: Failed to retrieve type information for service '/qc_camera2/set_exposure' of type 'camera_control_msgs/SetExposure'
2023-09-01T21:43:53.922878037Z [ WARN] [1693604633.922837389]: Could not find package 'camera_control_msgs' in srv paths!
[ERROR] [1693604633.922858760]: Failed to retrieve type information for service '/qc_camera2/set_gain' of type 'camera_control_msgs/SetGain'
[ WARN] [1693604633.928546769]: Could not find package 'camera_control_msgs' in srv paths!
[ERROR] [1693604633.928573579]: Failed to retrieve type information for service '/qc_camera2/set_gamma' of type 'camera_control_msgs/SetGamma'
[ WARN] [1693604633.931902503]: Could not find package 'camera_control_msgs' in srv paths!
[ERROR] [1693604633.931927521]: Failed to retrieve type information for service '/qc_camera2/get_cam_properties' of type 'camera_control_msgs/GetCamProperties'
[ WARN] [1693604633.934509107]: Could not find package 'arena_camera_ros' in srv paths!
[ERROR] [1693604633.934533724]: Failed to retrieve type information for service '/qc_camera2/toggle_stream' of type 'arena_camera_ros/ToggleCameraStream'
[ WARN] [1693604634.000754093]: Could not find package 'camera_control_msgs' in srv paths!
[ERROR] [1693604634.000793578]: Failed to retrieve type information for service '/qc_camera3/set_binning' of type 'camera_control_msgs/SetBinning'
[ WARN] [1693604634.006043265]: Could not find package 'camera_control_msgs' in srv paths!
[ERROR] [1693604634.006080195]: Failed to retrieve type information for service '/qc_camera3/set_exposure' of type 'camera_control_msgs/SetExposure'
2023-09-01T21:43:54.010520507Z [ WARN] [1693604634.010460173]: Could not find package 'camera_control_msgs' in srv paths!
[ERROR] [1693604634.010488476]: Failed to retrieve type information for service '/qc_camera3/set_gain' of type 'camera_control_msgs/SetGain'
[ WARN] [1693604634.012819628]: Could not find package 'camera_control_msgs' in srv paths!
[ERROR] [1693604634.012843794]: Failed to retrieve type information for service '/qc_camera3/set_gamma' of type 'camera_control_msgs/SetGamma'
2023-09-01T21:43:54.015401124Z [ WARN] [1693604634.015358714]: Could not find package 'camera_control_msgs' in srv paths!
[ERROR] [1693604634.015384623]: Failed to retrieve type information for service '/qc_camera3/get_cam_properties' of type 'camera_control_msgs/GetCamProperties'
2023-09-01T21:43:54.017577041Z [ WARN] [1693604634.017520243]: Could not find package 'arena_camera_ros' in srv paths!
[ERROR] [1693604634.017543197]: Failed to retrieve type information for service '/qc_camera3/toggle_stream' of type 'arena_camera_ros/ToggleCameraStream'
[ WARN] [1693604634.888113285]: Failed to retrieve type of service /rostopic_3080030_1693521791615/get_loggers
2023-09-01T21:43:55.162104952Z [ WARN] [1693604635.161994142]: Failed to retrieve type of service /rqt_gui_cpp_node_54672/get_loggers
[ WARN] [1693604635.414474287]: Failed to retrieve type of service /rqt_gui_cpp_node_54672/set_logger_level
2023-09-01T21:43:55.493921674Z [ERROR] [1693604635.493826935]: Failed to retrieve service type or service description of service /azure1/k4a_nodelet_manager/get_loggers: std::future_error: Broken promise
[ERROR] [1693604635.495109848]: Failed to retrieve service type or service description of service /azure1/k4a_nodelet_manager/set_logger_level: std::future_error: Broken promise
[ERROR] [1693604635.496578403]: Failed to retrieve service type or service description of service /azure1/k4a_nodelet_manager/load_nodelet: std::future_error: Broken promise
[ERROR] [1693604635.497926219]: Failed to retrieve service type or service description of service /azure1/k4a_nodelet_manager/unload_nodelet: std::future_error: Broken promise
2023-09-01T21:43:55.499340081Z [ERROR] [1693604635.499303221]: Failed to retrieve service type or service description of service /azure1/k4a_nodelet_manager/list: std::future_error: Broken promise
[ERROR] [1693604635.500588469]: Failed to retrieve service type or service description of service /robot_2_dx_motion_interface/get_loggers: std::future_error: Broken promise
[ WARN] [1693604635.751974429]: Failed to retrieve type of service /robot_2_dx_motion_interface/set_logger_level
[ERROR] [1693604635.754514737]: Failed to retrieve service type or service description of service /robot_2_fanuc_rmi_client/get_loggers: std::future_error: Broken promise
[ERROR] [1693604635.759561830]: Failed to retrieve service type or service description of service /robot_2_fanuc_rmi_client/set_logger_level: std::future_error: Broken promise
2023-09-01T21:43:55.765559225Z [ERROR] [1693604635.765512447]: Failed to retrieve service type or service description of service /R2_VeneerCornerObserver/get_loggers: std::future_error: Broken promise
[ERROR] [1693604635.766427602]: Failed to retrieve service type or service description of service /R2_VeneerCornerObserver/set_logger_level: std::future_error: Broken promise
[ERROR] [1693604635.767290950]: Failed to retrieve service type or service description of service /robot_1_dx_motion_interface/get_loggers: std::future_error: Broken promise
[ERROR] [1693604635.768171179]: Failed to retrieve service type or service description of service /robot_1_dx_motion_interface/set_logger_level: std::future_error: Broken promise
[ERROR] [1693604635.769066327]: Failed to retrieve service type or service description of service /robot_1_fanuc_rmi_client/get_loggers: std::future_error: Broken promise
[ERROR] [1693604635.773180531]: Failed to retrieve service type or service description of service /robot_1_fanuc_rmi_client/set_logger_level: std::future_error: Broken promise
[ERROR] [1693604635.924435023]: Failed to retrieve service type or service description of service /rviz_msiraider_455523_4051231953047898198/get_loggers: std::future_error: Broken promise
[ERROR] [1693604635.930194718]: Failed to retrieve service type or service description of service /rviz_msiraider_455523_4051231953047898198/set_logger_level: std::future_error: Broken promise
[ERROR] [1693604635.935104139]: Failed to retrieve service type or service description of service /rviz_msiraider_455523_4051231953047898198/reload_shaders: std::future_error: Broken promise
[ERROR] [1693604635.939061535]: Failed to retrieve service type or service description of service /rviz_msiraider_455523_4051231953047898198/load_config: std::future_error: Broken promise
[ERROR] [1693604635.942793704]: Failed to retrieve service type or service description of service /rviz_msiraider_455523_4051231953047898198/load_config_discarding_changes: std::future_error: Broken promise
2023-09-01T21:43:55.946994611Z [ERROR] [1693604635.946937263]: Failed to retrieve service type or service description of service /rviz_msiraider_455523_4051231953047898198/save_config: std::future_error: Broken promise
[ERROR] [1693604635.948405397]: Failed to retrieve service type or service description of service /rqt_gui_cpp_node_3355783/get_loggers: std::future_error: Broken promise
2023-09-01T21:43:55.949828576Z [ERROR] [1693604635.949785685]: Failed to retrieve service type or service description of service /rqt_gui_cpp_node_3355783/set_logger_level: std::future_error: Broken promise
[ERROR] [1693604635.952033509]: Failed to retrieve service type or service description of service /rqt_gui_cpp_node_3360882/get_loggers: std::future_error: Broken promise
[ERROR] [1693604635.953121563]: Failed to retrieve service type or service description of service /rqt_gui_cpp_node_3360882/set_logger_level: std::future_error: Broken promise
2023-09-01T21:43:55.955427967Z [ERROR] [1693604635.955358255]: Failed to retrieve service type or service description of service /tf_echo_1693005407704472780/get_loggers: std::future_error: Broken promise
[ERROR] [1693604635.958549357]: Failed to retrieve service type or service description of service /tf_echo_1693005407704472780/set_logger_level: std::future_error: Broken promise
[ERROR] [1693604635.963378256]: Failed to retrieve service type or service description of service /rviz_msiraider_475863_1805560526383919607/get_loggers: std::future_error: Broken promise
[ERROR] [1693604635.967455169]: Failed to retrieve service type or service description of service /rviz_msiraider_475863_1805560526383919607/set_logger_level: std::future_error: Broken promise
[ERROR] [1693604635.973375127]: Failed to retrieve service type or service description of service /rviz_msiraider_475863_1805560526383919607/reload_shaders: std::future_error: Broken promise
[ERROR] [1693604635.978929322]: Failed to retrieve service type or service description of service /rviz_msiraider_475863_1805560526383919607/load_config: std::future_error: Broken promise
[ERROR] [1693604635.983714979]: Failed to retrieve service type or service description of service /rviz_msiraider_475863_1805560526383919607/load_config_discarding_changes: std::future_error: Broken promise
[ERROR] [1693604635.988177192]: Failed to retrieve service type or service description of service /rviz_msiraider_475863_1805560526383919607/save_config: std::future_error: Broken promise
[ WARN] [1693604636.240285112]: Failed to retrieve type of service /rqt_gui_cpp_node_68845/get_loggers
[ERROR] [1693604636.245580837]: Failed to retrieve service type or service description of service /rqt_gui_cpp_node_68845/set_logger_level: std::future_error: Broken promise
[ WARN] [1693604636.497235276]: Failed to retrieve type of service /rqt_gui_cpp_node_11371/get_loggers
[ WARN] [1693604636.750249946]: Failed to retrieve type of service /rqt_gui_cpp_node_11371/set_logger_level
[ WARN] [1693604637.002703832]: Failed to retrieve type of service /rqt_gui_cpp_node_15179/get_loggers
[ WARN] [1693604637.255063360]: Failed to retrieve type of service /rqt_gui_cpp_node_15179/set_logger_level
[ WARN] [1693604637.507601486]: Failed to retrieve type of service /rqt_gui_cpp_node_15179/compressed/set_parameters
[ERROR] [1693604637.510493360]: Failed to retrieve service type or service description of service /tf_echo_1693329846940349580/get_loggers: std::future_error: Broken promise
[ERROR] [1693604637.512003444]: Failed to retrieve service type or service description of service /tf_echo_1693329846940349580/set_logger_level: std::future_error: Broken promise
[ERROR] [1693604637.515548539]: Failed to retrieve service type or service description of service /rviz_msiraider_613058_4222886344543863053/get_loggers: std::future_error: Broken promise
[ERROR] [1693604637.518957915]: Failed to retrieve service type or service description of service /rviz_msiraider_613058_4222886344543863053/set_logger_level: std::future_error: Broken promise
[ERROR] [1693604637.523424577]: Failed to retrieve service type or service description of service /rviz_msiraider_613058_4222886344543863053/reload_shaders: std::future_error: Broken promise
[ERROR] [1693604637.526815428]: Failed to retrieve service type or service description of service /rviz_msiraider_613058_4222886344543863053/load_config: std::future_error: Broken promise
[ERROR] [1693604637.530761303]: Failed to retrieve service type or service description of service /rviz_msiraider_613058_4222886344543863053/load_config_discarding_changes: std::future_error: Broken promise
[ERROR] [1693604637.534517817]: Failed to retrieve service type or service description of service /rviz_msiraider_613058_4222886344543863053/save_config: std::future_error: Broken promise
[ERROR] [1693604637.537352724]: Failed to retrieve service type or service description of service /rostopic_614368_1693332654733/get_loggers: std::future_error: Broken promise
[ERROR] [1693604642.534917374]: Failed to retrieve service type or service description of service /tf_echo_1693329846940349580/get_loggers: std::future_error: Broken promise
[ERROR] [1693604642.535978898]: Failed to retrieve service type or service description of service /tf_echo_1693329846940349580/set_logger_level: std::future_error: Broken promise
[ERROR] [1693604642.539597401]: Failed to retrieve service type or service description of service /rviz_msiraider_613058_4222886344543863053/get_loggers: std::future_error: Broken promise
[ERROR] [1693604642.542813721]: Failed to retrieve service type or service description of service /rviz_msiraider_613058_4222886344543863053/set_logger_level: std::future_error: Broken promise
[ERROR] [1693604642.546250369]: Failed to retrieve service type or service description of service /rviz_msiraider_613058_4222886344543863053/reload_shaders: std::future_error: Broken promise
[ERROR] [1693604642.549297780]: Failed to retrieve service type or service description of service /rviz_msiraider_613058_4222886344543863053/load_config: std::future_error: Broken promise
[ERROR] [1693604642.553177488]: Failed to retrieve service type or service description of service /rviz_msiraider_613058_4222886344543863053/load_config_discarding_changes: std::future_error: Broken promise
[ERROR] [1693604642.556038374]: Failed to retrieve service type or service description of service /rviz_msiraider_613058_4222886344543863053/save_config: std::future_error: Broken promise
[ERROR] [1693604642.558357262]: Failed to retrieve service type or service description of service /rostopic_614368_1693332654733/get_loggers: std::future_error: Broken promise
[ERROR] [1693604642.560864648]: Failed to retrieve service type or service description of service /rostopic_614368_1693332654733/set_logger_level: std::future_error: Broken promise
[ERROR] [1693604642.564076470]: Failed to retrieve service type or service description of service /rviz_msiraider_618543_7783352244266021340/get_loggers: std::future_error: Broken promise
[ERROR] [1693604642.567137526]: Failed to retrieve service type or service description of service /rviz_msiraider_618543_7783352244266021340/set_logger_level: std::future_error: Broken promise
[ERROR] [1693604642.569462505]: Failed to retrieve service type or service description of service /rviz_msiraider_618543_7783352244266021340/reload_shaders: std::future_error: Broken promise
[ERROR] [1693604642.572294607]: Failed to retrieve service type or service description of service /rviz_msiraider_618543_7783352244266021340/load_config: std::future_error: Broken promise
[ERROR] [1693604642.575269139]: Failed to retrieve service type or service description of service /rviz_msiraider_618543_7783352244266021340/load_config_discarding_changes: std::future_error: Broken promise
[ERROR] [1693604642.578592703]: Failed to retrieve service type or service description of service /rviz_msiraider_618543_7783352244266021340/save_config: std::future_error: Broken promise
[ WARN] [1693604642.829554920]: Failed to retrieve type of service /rqt_gui_cpp_node_19923/get_loggers
[ WARN] [1693604643.082882946]: Failed to retrieve type of service /rqt_gui_cpp_node_19923/set_logger_level
[ WARN] [1693604643.335065919]: Failed to retrieve type of service /rqt_gui_cpp_node_11730/get_loggers
[ WARN] [1693604643.587460093]: Failed to retrieve type of service /rqt_gui_cpp_node_11730/set_logger_level
[ WARN] [1693604643.839907638]: Failed to retrieve type of service /rqt_gui_cpp_node_11730/compressed/set_parameters
[ERROR] [1693604643.845982721]: Failed to retrieve service type or service description of service /LogCommanderTester_cpp/get_loggers: std::future_error: Broken promise
[ERROR] [1693604643.850509587]: Failed to retrieve service type or service description of service /LogCommanderTester_cpp/set_logger_level: std::future_error: Broken promise
[ERROR] [1693604643.853895659]: Failed to retrieve service type or service description of service /rviz_msiraider_712938_3675954756021939125/get_loggers: std::future_error: Broken promise
[ERROR] [1693604643.857126998]: Failed to retrieve service type or service description of service /rviz_msiraider_712938_3675954756021939125/set_logger_level: std::future_error: Broken promise
[ERROR] [1693604643.861228538]: Failed to retrieve service type or service description of service /rviz_msiraider_712938_3675954756021939125/reload_shaders: std::future_error: Broken promise
[ERROR] [1693604643.865680351]: Failed to retrieve service type or service description of service /rviz_msiraider_712938_3675954756021939125/load_config: std::future_error: Broken promise
[ERROR] [1693604643.870866328]: Failed to retrieve service type or service description of service /rviz_msiraider_712938_3675954756021939125/load_config_discarding_changes: std::future_error: Broken promise
[ERROR] [1693604643.874622261]: Failed to retrieve service type or service description of service /rviz_msiraider_712938_3675954756021939125/save_config: std::future_error: Broken promise
[ERROR] [1693604643.994334548]: Failed to retrieve service type or service description of service /rviz_1693500241430558177/get_loggers: std::future_error: Broken promise
[ERROR] [1693604644.007334375]: Failed to retrieve service type or service description of service /rviz_1693500241430558177/set_logger_level: std::future_error: Broken promise
[ERROR] [1693604644.026646084]: Failed to retrieve service type or service description of service /rviz_1693500241430558177/reload_shaders: std::future_error: Broken promise
[ERROR] [1693604644.053180772]: Failed to retrieve service type or service description of service /rviz_1693500241430558177/load_config: std::future_error: Broken promise
[ERROR] [1693604644.172584975]: Failed to retrieve service type or service description of service /rviz_1693500241430558177/load_config_discarding_changes: std::future_error: Broken promise
[ERROR] [1693604644.290143928]: Failed to retrieve service type or service description of service /rviz_1693500241430558177/save_config: std::future_error: Broken promise
[ WARN] [1693604644.542250256]: Failed to retrieve type of service /rviz_1693500241430558177/planningscene_planning_scene_monitor/set_parameters
[ WARN] [1693604644.794886449]: Failed to retrieve type of service /roboflow/get_loggers
[ WARN] [1693604645.047990019]: Failed to retrieve type of service /roboflow/set_logger_level
[ERROR] [1693604645.280006653]: Failed to retrieve service type or service description of service /rqt_gui_cpp_node_14210/get_loggers: std::future_error: Broken promise
[ WARN] [1693604645.532430263]: Failed to retrieve type of service /rqt_gui_cpp_node_14210/set_logger_level
[ WARN] [1693604645.785340617]: Failed to retrieve type of service /rqt_gui_cpp_node_54672/get_loggers
[ WARN] [1693604646.038029360]: Failed to retrieve type of service /rqt_gui_cpp_node_54672/set_logger_level
[ERROR] [1693604646.040248199]: Failed to retrieve service type or service description of service /azure1/k4a_nodelet_manager/get_loggers: std::future_error: Broken promise
[ERROR] [1693604646.041587869]: Failed to retrieve service type or service description of service /azure1/k4a_nodelet_manager/set_logger_level: std::future_error: Broken promise
[ERROR] [1693604646.042808514]: Failed to retrieve service type or service description of service /azure1/k4a_nodelet_manager/load_nodelet: std::future_error: Broken promise
[ERROR] [1693604646.043979766]: Failed to retrieve service type or service description of service /azure1/k4a_nodelet_manager/unload_nodelet: std::future_error: Broken promise
[ERROR] [1693604646.045178359]: Failed to retrieve service type or service description of service /azure1/k4a_nodelet_manager/list: std::future_error: Broken promise
2023-09-01T21:44:06.046421617Z [ERROR] [1693604646.046374779]: Failed to retrieve service type or service description of service /robot_2_dx_motion_interface/get_loggers: std::future_error: Broken promise
[ERROR] [1693604646.047603399]: Failed to retrieve service type or service description of service /robot_2_dx_motion_interface/set_logger_level: std::future_error: Broken promise
2023-09-01T21:44:06.048841758Z [ERROR] [1693604646.048795140]: Failed to retrieve service type or service description of service /robot_2_fanuc_rmi_client/get_loggers: std::future_error: Broken promise
[ERROR] [1693604646.049998602]: Failed to retrieve service type or service description of service /robot_2_fanuc_rmi_client/set_logger_level: std::future_error: Broken promise
[ERROR] [1693604646.051181506]: Failed to retrieve service type or service description of service /R2_VeneerCornerObserver/get_loggers: std::future_error: Broken promise
[ERROR] [1693604646.052349872]: Failed to retrieve service type or service description of service /R2_VeneerCornerObserver/set_logger_level: std::future_error: Broken promise
[ERROR] [1693604646.053398030]: Failed to retrieve service type or service description of service /robot_1_dx_motion_interface/get_loggers: std::future_error: Broken promise
[ERROR] [1693604646.054187878]: Failed to retrieve service type or service description of service /robot_1_dx_motion_interface/set_logger_level: std::future_error: Broken promise
[ERROR] [1693604646.054961996]: Failed to retrieve service type or service description of service /robot_1_fanuc_rmi_client/get_loggers: std::future_error: Broken promise
[ERROR] [1693604646.055747836]: Failed to retrieve service type or service description of service /robot_1_fanuc_rmi_client/set_logger_level: std::future_error: Broken promise
[ERROR] [1693604646.132841009]: [WS] handle_accept error: Operation aborted.
[ INFO] [1693604629.821394625]: Loading nodelet /foxglove_bridge of type foxglove_bridge/foxglove_bridge_nodelet to manager foxglove_nodelet_manager with the following remappings:
2023-09-01T21:44:08.961000753Z [ INFO] [1693604629.834298210]: waitForService: Service [/foxglove_nodelet_manager/load_nodelet] has not been advertised, waiting...
[ INFO] [1693604629.856897093]: waitForService: Service [/foxglove_nodelet_manager/load_nodelet] is now available.
[ INFO] [1693604648.840659673]: Bond broken, exiting
``` 

- Version: 0.7.1
- Platform: Ubuntu 20:04 - ROS1 Noetic

**Steps To Reproduce**
Launch foxglove-bridge along side instances of RVIZ that are hosted on second machine. 


**Expected Behavior**
Exception is that the node continues on even if its unable to get specific services.

Looks like its not fully fixed. We get another broken bond when we terminate a foxglove studio instance connected to this patched bridge. 


```
2023-09-05T16:49:55.569952054Z [ INFO] [1693932595.567760045]: [WS] Unsubscribing from parameter '/oak2/stereo_i_low_bandwidth'.
[ INFO] [1693932595.567765616]: [WS] Unsubscribing from parameter '/oak2/stereo_i_low_bandwidth_quality'.
[ INFO] [1693932595.567771116]: [WS] Unsubscribing from parameter '/oak2/stereo_i_spatial_filter_alpha'.
[ERROR] [1693932595.782057741]: Failed to retrieve service type or service description of service /rviz_msiraider_712938_3675954756021939125/reload_shaders: Failed to retrieve service type
[ INFO] [1693932595.567776657]: [WS] Unsubscribing from parameter '/oak2/stereo_i_spatial_filter[ INFO] [1693932555.211025287]: Loading nodelet /foxglove_bridge of type foxglove_bridge/foxglove_bridge_nodelet to manager foxglove_nodelet_manager with the following remappings:
2023-09-05T16:49:55.958292364Z [ INFO] [1693932555.230931624]: waitForService: Service [/foxglove_nodelet_manager/load_nodelet] has not been advertised, waiting...
[ INFO] [1693932555.253678450]: waitForService: Service [/foxglove_nodelet_manager/load_nodelet] is now available.
2023-09-05T16:49:55.958318914Z [ INFO] [1693932595.739912453]: Bond broken, exiting
2023-09-05T16:49:56.002073967Z [foxglove_bridge-2] process has finished cleanly
```

Trying with services disabled, we get:

```
master has started, initiating launch
... logging to /root/.ros/log/rios_1693934935/roslaunch-hood1-1-headnode-1-1.log
Checking log directory for disk usage. This may take a while.
Press Ctrl-C to interrupt
Done checking log file disk usage. Usage is <1GB.
]2;/opt/rios/install/share/foxglove_bridge/launch/foxglove_bridge.launch
started roslaunch server http://hood1-1-headnode-1:40387/

SUMMARY
========
2023-09-05T17:58:56.277362044Z 
PARAMETERS
2023-09-05T17:58:56.277368045Z  * /foxglove_bridge/address: 0.0.0.0
 * /foxglove_bridge/asset_uri_allowlist: ['package://(/?\\...
 * /foxglove_bridge/capabilities: ['clientPublish',...
 * /foxglove_bridge/certfile: 
 * /foxglove_bridge/client_topic_whitelist: ['.*']
 * /foxglove_bridge/keyfile: 
2023-09-05T17:58:56.277387201Z  * /foxglove_bridge/max_update_ms: 5000
 * /foxglove_bridge/param_whitelist: ['.*']
 * /foxglove_bridge/port: 8765
2023-09-05T17:58:56.277396048Z  * /foxglove_bridge/send_buffer_limit: 10000000
 * /foxglove_bridge/service_whitelist: ['.*']
2023-09-05T17:58:56.277401980Z  * /foxglove_bridge/tls: False
 * /foxglove_bridge/topic_whitelist: ['.*']
2023-09-05T17:58:56.277407831Z  * /foxglove_nodelet_manager/num_worker_threads: 0
 * /rosdistro: noetic
2023-09-05T17:58:56.277413742Z  * /rosversion: 1.16.0

NODES
  /
2023-09-05T17:58:56.277451143Z     foxglove_bridge (nodelet/nodelet)
    foxglove_nodelet_manager (nodelet/nodelet)

ROS_MASTER_URI=http://hood1-1-rios-roscore-service:80/
2023-09-05T17:58:56.281547796Z ]2;/opt/rios/install/share/foxglove_bridge/launch/foxglove_bridge.launch http://hood1-1-rios-roscore-service:80/
2023-09-05T17:58:56.323674728Z process[foxglove_nodelet_manager-1]: started with pid [68]
process[foxglove_bridge-2]: started with pid [69]
[ WARN] [1693936736.417956380]: Could not find package 'control_msgs' in msg paths!
2023-09-05T17:58:56.418076828Z [ WARN] [1693936736.417983291]: Could not find definition for type control_msgs/FollowJointTrajectoryActionResult
[ WARN] [1693936736.418470826]: Could not find package 'control_msgs' in msg paths!
[ WARN] [1693936736.418482598]: Could not find definition for type control_msgs/FollowJointTrajectoryActionResult
2023-09-05T17:58:56.418651048Z [ WARN] [1693936736.418621492]: Could not find package 'control_msgs' in msg paths!
2023-09-05T17:58:56.418654064Z [ WARN] [1693936736.418631801]: Could not find definition for type control_msgs/FollowJointTrajectoryActionFeedback
2023-09-05T17:58:56.418686976Z [ WARN] [1693936736.418658913]: Could not find package 'control_msgs' in msg paths!
2023-09-05T17:58:56.418691395Z [ WARN] [1693936736.418667650]: Could not find definition for type control_msgs/FollowJointTrajectoryActionFeedback
2023-09-05T17:58:56.418897496Z [ WARN] [1693936736.418860716]: Could not find package 'dx_observer_msgs' in msg paths!
[ WARN] [1693936736.418870945]: Could not find definition for type dx_observer_msgs/Observation
[ERROR] [1693936746.521918680]: [WS] handle_accept error: Operation aborted.
[ INFO] [1693936736.357754332]: Initializing nodelet with 64 worker threads.
[ INFO] [1693936736.408834302]: Starting foxglove_bridge (noetic, 0.7.1@53ad89b) with WebSocket++/0.8.1
2023-09-05T17:59:06.527656057Z [ INFO] [1693936736.409003203]: [WS] Server running without TLS
[ INFO] [1693936736.409237186]: [WS] WebSocket server listening at ws://0.0.0.0:8765
2023-09-05T17:59:06.527674572Z [ INFO] [1693936736.409344370]: [WS] WebSocket server run loop started
[ INFO] [1693936739.164181210]: [WS] Client 100.123.190.98:40170 connected via /
2023-09-05T17:59:06.527687306Z [ INFO] [1693936739.517243286]: [WS] Subscribing to connection graph updates.
[ INFO] [1693936739.518012145]: Client 100.123.190.98:40170 is advertising "/move_base_simple/goal" (geometry_msgs/PoseStamped) on channel 1
[ INFO] [1693936739.518422734]: Client 100.123.190.98:40170 is advertising "/clicked_point" (geometry_msgs/PointStamped) on channel 2
[ INFO] [1693936739.518816842]: Client 100.123.190.98:40170 is advertising "/initialpose" (geometry_msgs/PoseWithCovarianceStamped) on channel 3
[ INFO] [1693936739.826677673]: [WS] Client 10.100.0.106:57592 connected via /
2023-09-05T17:59:06.527735347Z [ INFO] [1693936740.160383853]: Client 10.100.0.106:57592 is advertising "/clicked_point" (geometry_msgs/PointStamped) on channel 1
[ INFO] [1693936740.160431323]: Client 10.100.0.106:57592 is advertising "/initialpose" (geometry_msgs/PoseWithCovarianceStamped) on channel 2
2023-09-05T17:59:06.527748141Z [ INFO] [1693936740.160459857]: Client 10.100.0.106:57592 is no longer advertising /clicked_point (0 subscribers) on channel 1
[ INFO] [1693936740.160491778]: Client 10.100.0.106:57592 is no longer advertising /initialpose (0 subscribers) on channel 2
[ INFO] [1693936740.160518518]: Client 10.100.0.106:57592 is advertising "/move_base_simple/goal" (geometry_msgs/PoseStamped) on channel 3
[ INFO] [1693936740.160537134]: Client 10.100.0.106:57592 is advertising "/clicked_point" (geometry_msgs/PointStamped) on channel 4
2023-09-05T17:59:06.527773119Z [ INFO] [1693936740.160559296]: Client 10.100.0.106:57592 is advertising "/initialpose" (geometry_msgs/PoseWithCovarianceStamped) on channel 5
[ INFO] [1693936740.162863007]: Subscribed to topic "/tf" (tf2_msgs/TFMessage) on channel 349
[ INFO] [1693936740.166149783]: Subscribed to topic "/tf_static" (tf2_msgs/TFMessage) on channel 347
[ INFO] [1693936740.606902956]: Client 100.123.190.98:40170 is no longer advertising /move_base_simple/goal (0 subscribers) on channel 1
2023-09-05T17:59:06.527796724Z [ INFO] [1693936740.606948613]: Client 100.123.190.98:40170 is no longer advertising /clicked_point (0 subscribers) on channel 2
[ INFO] [1693936740.606973540]: Client 100.123.190.98:40170 is no longer advertising /initialpose (0 subscribers) on channel 3
2023-09-05T17:59:06.527808376Z [ INFO] [1693936740.607031139]: Client 100.123.190.98:40170 is advertising "/move_base_simple/goal" (geometry_msgs/PoseStamped) on channel 4
[ INFO] [1693936740.607068971]: Client 100.123.190.98:40170 is advertising "/clicked_point" (geometry_msgs/PointStamped) on channel 5
2023-09-05T17:59:06.527820248Z [ INFO] [1693936740.607105240]: Client 100.123.190.98:40170 is advertising "/initialpose" (geometry_msgs/PoseWithCovarianceStamped) on channel 6
[ INFO] [1693936740.650652567]: Added subscriber #2 to topic "/tf" (tf2_msgs/TFMessage) on channel 349
2023-09-05T17:59:06.527831580Z [ INFO] [1693936740.652242534]: Added subscriber #2 to topic "/tf_static" (tf2_msgs/TFMessage) on channel 347
[ INFO] [1693936746.521687662]: [WS] Stopping WebSocket server
2023-09-05T17:59:06.527842741Z [ INFO] [1693936746.521761091]: [WS] Closing 2 client connection(s)
[ INFO] [1693936746.521796058]: [WS] asio handle_accept error: asio.system:125 (Operation aborted.)
[ INFO] [1693936746.521846954]: [WS] Error getting remote endpoint: asio.system:9 (Bad file descriptor)
[ INFO] [1693936746.521886690]: [WS] asio async_shutdown error: asio.system:9 (Bad file descriptor)
[ INFO] [1693936746.521968886]: [WS] Stopping acceptance of new connections because the underlying transport is no longer listening.
[ INFO] [1693936746.527498498]: [WS] Client 10.100.0.106:57592 disconnected
[ INFO] [1693936746.527552019]: [WS] Client 10.100.0.106:57592 unadvertising channel 5 due to disconnect
2023-09-05T17:59:06.535630758Z [ INFO] [1693936746.527584441]: Client 10.100.0.106:57592 is no longer advertising /initialpose (0 subscribers) on channel 5
2023-09-05T17:59:06.535635066Z [ INFO] [1693936746.527603617]: [WS] Client 10.100.0.106:57592 unadvertising channel 4 due to disconnect
[ INFO] [1693936746.527620159]: Client 10.100.0.106:57592 is no longer advertising /clicked_point (0 subscribers) on channel 4
2023-09-05T17:59:06.535644394Z [ INFO] [1693936746.527633524]: [WS] Client 10.100.0.106:57592 unadvertising channel 3 due to disconnect
[ INFO] [1693936746.527655506]: Client 10.100.0.106:57592 is no longer advertising /move_base_simple/goal (0 subscribers) on channel 3
[ INFO] [1693936746.527703477]: Removed one subscription from channel 347 (1 subscription(s) left)
[ INFO] [1693936746.527718195]: Removed one subscription from channel 349 (1 subscription(s) left)
2023-09-05T17:59:06.535663410Z [ INFO] [1693936746.531452430]: [WS] WebSocket server run loop stopped
[ INFO] [1693936746.531523405]: [WS] Client 100.123.190.98:40170 disconnected
[ INFO] [1693936746.531545657]: [WS] Client 100.123.190.98:40170 unadvertising channel 6 due to disconnect
[ INFO] [1693936746.531571185]: Client 100.123.190.98:40170 is no longer advertising /initialpose (0 subscribers) on channel 6
[ INFO] [1693936746.531868309]: [WS] All WebSocket connections closed
[ INFO] [1693936746.531891253]: [WS] Waiting for WebSocket server run loop to terminate
[ INFO] [1693936746.531907223]: [WS] WebSocket server run loop terminated
[ INFO] [1693936746.532393366]: [WS] Client 100.123.190.98:40170 unadvertising channel 5 due to disconnect
[ INFO] [1693936746.532412712]: Client (unknown) is no longer advertising /clicked_point (0 subscribers) on channel 5
[ INFO] [1693936746.532865832]: [WS] Client 100.123.190.98:40170 unadvertising channel 4 due to disconnect
[ INFO] [1693936746.532887112]: Client (unknown) is no longer advertising /move_base_simple/goal (0 subscribers) on channel 4
[ INFO] [1693936746.533980327]: Unsubscribing from topic "/tf_static" (tf2_msgs/TFMessage) on channel 347
2023-09-05T17:59:06.535732862Z [ INFO] [1693936746.534512356]: Unsubscribing from topic "/tf" (tf2_msgs/TFMessage) on channel 349

LOTS OF UNSUBSCRIBING...

[ INFO] [1693936736.353239575]: waitForService: Service [/foxglove_nodelet_manager/load_nodelet] has not been advertised, waiting...
2023-09-05T17:59:09.816259286Z [ INFO] [1693936736.374996878]: waitForService: Service [/foxglove_nodelet_manager/load_nodelet] is now available.
[ INFO] [1693936749.459915676]: Bond broken, exiting
```

--- a/ros1_foxglove_bridge/include/foxglove_bridge/service_utils.hpp
+++ b/ros1_foxglove_bridge/include/foxglove_bridge/service_utils.hpp
@@ -1,5 +1,6 @@
 #pragma once
 
+#include <chrono>
 #include <string>
 
 namespace foxglove_bridge {
@@ -8,6 +9,7 @@ namespace foxglove_bridge {
  * Opens a socket to the service server and retrieves the service type from the connection header.
  * This is necessary as the service type is not stored on the ROS master.
  */
-std::string retrieveServiceType(const std::string& serviceName, int timeout_ms);
+std::string retrieveServiceType(const std::string& serviceName,
+                                std::chrono::milliseconds timeout_ms);
 
 }  // namespace foxglove_bridge

--- a/ros1_foxglove_bridge/include/foxglove_bridge/service_utils.hpp
+++ b/ros1_foxglove_bridge/include/foxglove_bridge/service_utils.hpp
@@ -1,6 +1,5 @@
 #pragma once
 
-#include <future>
 #include <string>
 
 namespace foxglove_bridge {
@@ -9,6 +8,6 @@ namespace foxglove_bridge {
  * Opens a socket to the service server and retrieves the service type from the connection header.
  * This is necessary as the service type is not stored on the ROS master.
  */
-std::future<std::string> retrieveServiceType(const std::string& serviceName);
+std::string retrieveServiceType(const std::string& serviceName, int timeout_ms);
 
 }  // namespace foxglove_bridge

--- a/ros1_foxglove_bridge/src/ros1_foxglove_bridge_nodelet.cpp
+++ b/ros1_foxglove_bridge/src/ros1_foxglove_bridge_nodelet.cpp
@@ -605,7 +605,8 @@ private:
       }
 
       try {
-        const auto serviceType = retrieveServiceType(serviceName, _serviceRetrievalTimeoutMs);
+        const auto serviceType =
+          retrieveServiceType(serviceName, std::chrono::milliseconds(_serviceRetrievalTimeoutMs));
         const auto srvDescription = _rosTypeInfoProvider.getServiceDescription(serviceType);
 
         foxglove::ServiceWithoutId service;

--- a/ros1_foxglove_bridge/src/ros1_foxglove_bridge_nodelet.cpp
+++ b/ros1_foxglove_bridge/src/ros1_foxglove_bridge_nodelet.cpp
@@ -47,7 +47,7 @@ constexpr char ROS1_CHANNEL_ENCODING[] = "ros1";
 constexpr uint32_t SUBSCRIPTION_QUEUE_LENGTH = 10;
 constexpr double MIN_UPDATE_PERIOD_MS = 100.0;
 constexpr uint32_t PUBLICATION_QUEUE_LENGTH = 10;
-constexpr int SERVICE_TYPE_RETRIEVAL_TIMEOUT_MS = 250;
+constexpr int DEFAULT_SERVICE_TYPE_RETRIEVAL_TIMEOUT_MS = 250;
 
 using ConnectionHandle = websocketpp::connection_hdl;
 using TopicAndDatatype = std::pair<std::string, std::string>;
@@ -75,6 +75,8 @@ public:
     _capabilities = nhp.param<std::vector<std::string>>(
       "capabilities", std::vector<std::string>(foxglove::DEFAULT_CAPABILITIES.begin(),
                                                foxglove::DEFAULT_CAPABILITIES.end()));
+    _serviceRetrievalTimeoutMs = nhp.param<int>("service_type_retrieval_timeout_ms",
+                                                DEFAULT_SERVICE_TYPE_RETRIEVAL_TIMEOUT_MS);
 
     const auto topicWhitelistPatterns =
       nhp.param<std::vector<std::string>>("topic_whitelist", {".*"});
@@ -602,15 +604,8 @@ private:
         continue;  // Already advertised
       }
 
-      auto serviceTypeFuture = retrieveServiceType(serviceName);
-      if (serviceTypeFuture.wait_for(std::chrono::milliseconds(
-            SERVICE_TYPE_RETRIEVAL_TIMEOUT_MS)) != std::future_status::ready) {
-        ROS_WARN("Failed to retrieve type of service %s", serviceName.c_str());
-        continue;
-      }
-
       try {
-        const auto serviceType = serviceTypeFuture.get();
+        const auto serviceType = retrieveServiceType(serviceName, _serviceRetrievalTimeoutMs);
         const auto srvDescription = _rosTypeInfoProvider.getServiceDescription(serviceType);
 
         foxglove::ServiceWithoutId service;
@@ -919,6 +914,7 @@ private:
   ros::Subscriber _clockSubscription;
   bool _useSimTime = false;
   std::vector<std::string> _capabilities;
+  int _serviceRetrievalTimeoutMs = DEFAULT_SERVICE_TYPE_RETRIEVAL_TIMEOUT_MS;
   std::atomic<bool> _subscribeGraphUpdates = false;
   std::unique_ptr<foxglove::CallbackQueue> _fetchAssetQueue;
 };

--- a/ros1_foxglove_bridge/src/service_utils.cpp
+++ b/ros1_foxglove_bridge/src/service_utils.cpp
@@ -9,7 +9,7 @@
 
 namespace foxglove_bridge {
 
-std::string retrieveServiceType(const std::string& serviceName, int timeout_ms) {
+std::string retrieveServiceType(const std::string& serviceName, std::chrono::milliseconds timeout) {
   auto link = ros::ServiceManager::instance()->createServiceServerLink(serviceName, false, "*", "*",
                                                                        {{"probe", "1"}});
   std::promise<std::string> promise;
@@ -27,8 +27,8 @@ std::string retrieveServiceType(const std::string& serviceName, int timeout_ms) 
       return true;
     });
 
-  if (future.wait_for(std::chrono::milliseconds(timeout_ms)) != std::future_status::ready) {
-    throw std::runtime_error("Failed to retrieve service type");
+  if (future.wait_for(timeout) != std::future_status::ready) {
+    throw std::runtime_error("Timed out when retrieving service type");
   }
 
   return future.get();

--- a/ros1_foxglove_bridge/src/service_utils.cpp
+++ b/ros1_foxglove_bridge/src/service_utils.cpp
@@ -1,3 +1,6 @@
+#include <chrono>
+#include <future>
+
 #include <ros/connection.h>
 #include <ros/service_manager.h>
 #include <ros/service_server_link.h>
@@ -6,25 +9,29 @@
 
 namespace foxglove_bridge {
 
-std::future<std::string> retrieveServiceType(const std::string& serviceName) {
+std::string retrieveServiceType(const std::string& serviceName, int timeout_ms) {
   auto link = ros::ServiceManager::instance()->createServiceServerLink(serviceName, false, "*", "*",
                                                                        {{"probe", "1"}});
-  auto promise = std::make_shared<std::promise<std::string>>();
-  auto future = promise->get_future();
+  std::promise<std::string> promise;
+  auto future = promise.get_future();
 
   link->getConnection()->setHeaderReceivedCallback(
-    [promise = std::move(promise)](const ros::ConnectionPtr&, const ros::Header& header) mutable {
+    [&promise](const ros::ConnectionPtr&, const ros::Header& header) {
       std::string serviceType;
       if (header.getValue("type", serviceType)) {
-        promise->set_value(serviceType);
+        promise.set_value(serviceType);
       } else {
-        promise->set_exception(std::make_exception_ptr(
+        promise.set_exception(std::make_exception_ptr(
           std::runtime_error("Key 'type' not found in service connection header")));
       }
       return true;
     });
 
-  return future;
+  if (future.wait_for(std::chrono::milliseconds(timeout_ms)) != std::future_status::ready) {
+    throw std::runtime_error("Failed to retrieve service type");
+  }
+
+  return future.get();
 }
 
 }  // namespace foxglove_bridge

